[Model Averaging] Provide a util function for model averaging

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60303 [Model Averaging] Provide a util function for model averaging**

The util function can be used for averaging parameters.

More optimizations can be done in the future.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

Commit:
[Model Averaging] Provide a util function for model averaging

The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -939,6 +940,31 @@ def test_new_subgroups_overlap_not_allowed(self):
+
+            # Every element should be the average of 0 and 1, i.e., 0.5.
+            for p in model.parameters():
+                self.assertEqual(p.data, torch.ones_like(p.data) * 0.5)
```
What if this test is ran on a 4-GPU machine, then ranks 2 and 3 would have 2/3s as model param data, and they shouldn't participate in the averaging process because the group passed in is only for ranks 0 and 1? Then it doesn't seem like this assert would succeed.

SciPioneer replied:
Yeah, there should have been a condition the current rank is in the subgroup.

Updated the test to test both global averaging and partial averaging.

rohan-varma commented on torch/distributed/algorithms/model_averaging/utils.py:
```diff
@@ -0,0 +1,20 @@
+    Thus, it requires extra memory of the same size as the model’s parameters.
+    """
+    flat_params = torch.cat([p.data.view(-1) for p in model.parameters()])
+    flat_params /= float(dist.get_world_size(group=process_group))
```
nit: I don't think we need the cast to float here?

SciPioneer replied:
Yeah, removed this unnecessary explicit casting.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -939,6 +940,31 @@ def test_new_subgroups_overlap_not_allowed(self):
                     ranks_per_subgroup_list=[[0], [1, 2], [1, 3]]
                 )
 
+        @require_backend({"nccl"})
```
I think we can enable this test for both gloo and nccl?

SciPioneer replied:
Yes, updated the decorator.

Commit:
Update on "[Model Averaging] Provide a util function for model averaging"


The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

Commit:
Update on "[Model Averaging] Provide a util function for model averaging"


The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

Commit:
Update on "[Model Averaging] Provide a util function for model averaging"


The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -939,6 +940,42 @@ def test_new_subgroups_overlap_not_allowed(self):
+                p.data = torch.ones_like(p.data) * rank
+            group_nccl = dist.new_group(ranks=[0, 1], backend="nccl")
+            model_averaging_utils.average_parameters(module=model, process_group=group_nccl)
+            if not dist._rank_not_in_group(group_nccl):
```
Should we enhance the test to verify that non-group members don't have their parameters touched?

SciPioneer replied:
Added

rohan-varma commented on torch/distributed/algorithms/model_averaging/utils.py:
```diff
@@ -0,0 +1,21 @@
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+
+    flat_params = torch.cat([p.data.view(-1) for p in module.parameters()])
+    flat_params /= dist.get_world_size(group_to_use)
```
If rank is not in group what happens here? For example, if we have a subgroup of [0, 1] would rank 2 divide their params by 2 and then not average them?

rohan-varma replied:
Actually I'm not sure if this will work properly if rank is not in group, as if rank is not in group and we call `get_world_size` we return `-1`: https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L824

So if the rank is not in group this might divide its parameters by -1? Can we test this?

SciPioneer replied:
Good question. The input of any rank out of the subgroup will be divided by `-1`, so the result of the input `rank` will be `-rank`.

rohan-varma reviewed (commented):
Looks good overall, just one concern about non-world subgroups.

Commit:
Update on "[Model Averaging] Provide a util function for model averaging"


The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -939,6 +940,47 @@ def test_new_subgroups_overlap_not_allowed(self):
+                # Every element on device not in the subgroup should be -rank,
+                # because `get_world_size()` returns -1.
+                for p in model.parameters():
+                    self.assertEqual(p.data, torch.ones_like(p.data) * -rank)
```
Ideally, non group members should not have their parameters modified right? can we just check `_rank_not_in_group` in the averaging function and early return if the rank isn't in group, so the params are not touched?

SciPioneer replied:
Yeah, no update should be better. An early return now is provided for non-group members.

Commit:
Update on "[Model Averaging] Provide a util function for model averaging"


The util function can be used for averaging parameters.

More optimizations can be done in the future.

Differential Revision: [D29242806](https://our.internmc.facebook.com/intern/diff/D29242806/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
lgtm
